src/drivers/snapdragon_pwm_out/snapdragon_pwm_out.cpp

```diff
@@ -0,0 +1,608 @@
+		}
+
+		return 0;
+
+	}
+
+}
+
+void subscribe()
+{
+	memset(_controls, 0, sizeof(_controls));
+	memset(_poll_fds, 0, sizeof(_poll_fds));
+
+	/*
+	 * ORB topic names to msg actuator_controls
+	 */
+	_controls_topics[0] = ORB_ID(actuator_controls_0);
+	_controls_topics[1] = ORB_ID(actuator_controls_1);
+	_controls_topics[2] = ORB_ID(actuator_controls_2);
+	_controls_topics[3] = ORB_ID(actuator_controls_3);
+
+
+	/*
+	 * subscibe to orb topic
+	 */
+	for (uint8_t i = 0; i < actuator_controls_s::NUM_ACTUATOR_CONTROL_GROUPS; i++) {
+
+		if (_groups_required & (1 << i)) {
+
+			PX4_DEBUG("subscribe to actuator_controls_%d", i);
+
+			_controls_subs[i] = orb_subscribe(_controls_topics[i]);
+
+		} else {
+
+			_controls_subs[i] = -1;
+
+		}
+
+		if (_controls_subs[i] >= 0) {
+
+			_poll_fds[_poll_fds_num].fd = _controls_subs[i];
+			_poll_fds[_poll_fds_num].events = POLLIN;
+			_poll_fds_num++;
+
+		}
+
+		_armed_sub = orb_subscribe(ORB_ID(actuator_armed));
+
+	}
+
+}
+
+
+
+int pwm_initialize(const char *device)
+{
+	/*
+	 * open PWM device
+	 */
+	int i = 0;
+	_fd = open("/dev/pwm-1", 0);
+
+	if (_fd < 0) {
+		PX4_ERR("failed to open PWM device!");
+		return -1;
+	}
+
+	/*
+	 * configure PWM
+	 */
+	for (i = 0; i < NUM_PWM; i++) {
+		_pwm_gpio[i].gpio_id = PIN_GPIO + i;
+	}
+
+	/*
+	 * description of signal
+	 */
+	_signal_definition.num_gpios = NUM_PWM;
+	_signal_definition.period_in_usecs = 1e6/FREQUENCY_PWM;
+	_signal_definition.pwm_signal = _pwm_gpio;
+
+
+	/*
+	 * send signal description to DSP
+	 */
+	if (::ioctl(_fd, PWM_IOCTL_SIGNAL_DEFINITION, &_signal_definition) != 0) {
+		PX4_ERR("failed to send signal to DSP");
+		return -1;
+	}
+
+	/*
+	 * retrive shared update buffer to make immediate changed in width of the pulse
+	 */
+	if (::ioctl(_fd, PWM_IOCTL_GET_UPDATE_BUFFER, &_update_buffer) != 0)
+	{
+		PX4_ERR("failed to receive update buffer ");
+		return -1;
+	}
+	_pwm = &_update_buffer->pwm_signal[0];
+
+
+	return 0;
+}
+
+void pwm_deinitialize()
+{
+	/*
+	 * close device ID
+	 */
+	close(_fd);
+
+}
+
+void send_outputs_pwm(const uint16_t *pwm)
+{
+
+
+	/*
+	 * send pwm in us: ToDo: check if it is in us
+	 */
+	for (unsigned i = 0; i < NUM_PWM; ++i) {
+		_pwm[i].pulse_width_in_usecs = pwm[i];
+	}
+}
+
+
+
+void task_main(int argc, char *argv[])
+{
+	_is_running = true;
+
+	if (pwm_initialize(_device) < 0) {
+		PX4_ERR("Failed to initialize PWM.");
+		return;
+	}
+
+	// Set up mixer
+	if (initialize_mixer(MIXER_FILENAME) < 0) {
+		PX4_ERR("Mixer initialization failed.");
+		return;
+	}
+
+	_mixer->groups_required(_groups_required);
+	// subscribe and set up polling
+	subscribe();
+
+	// Start disarmed
+	_armed.armed = false;
+	_armed.prearmed = false;
+
+	// set max min pwm
+	pwm_limit_init(&_pwm_limit);
+
+	// Main loop
+	while (!_task_should_exit) {
+
+		/* wait up to 10ms for data */
+		int pret = px4_poll(_poll_fds, _poll_fds_num, 10);
+
+
+		/* Timed out, do a periodic check for _task_should_exit. */
+		if (pret == 0) {
```

Comment:
I'm saying you should take some action if the timeout happens, e.g. setting PWM to the disarmed pulse.